### PR TITLE
Add ability to list S3 bucket contents

### PR DESCRIPTION
**Pull request by xemul**
This is prerequisite for "restore from object storage" feature. In order to collect the sstables in bucket one would need to list the bucket contents with the given prefix. The ListObjectsV2 provides a way for it and here's the respective s3::client extension.

### CODE REVIEW

File: utils/s3/client.cc [hunk 1]
```diff
@@ -1226,4 +1226,106 @@ future<> client::close() {
...
+
+        auto is_truncated = root_node->first_node("IsTruncated");
+        if (std::string_view(is_truncated->value()) == "false") {
+            return "";
+        } else {
+            auto continuation = root_node->first_node("NextContinuationToken");
+            return continuation->value();
+        }
+    } catch (...) {
+        s3l.warn("cannot parse initiate multipart upload response: {}", std::current_exception());
```
Comment (kreuzerkrieg) [inline utils/s3/client.cc:1257]: eh... "multipart upload response"? :)

Comment (xemul) [reply utils/s3/client.cc:1257]: :facepalm:

**Comment by xemul** [comment]
upd:
- Fix error message

File: test/boost/s3_test.cc [hunk 2]
```diff
@@ -377,3 +378,33 @@ SEASTAR_THREAD_TEST_CASE(test_client_put_get_tagging) {
...
+    auto close_client = deferred_close(*client);
+    std::unordered_set<sstring> names;
+
+    // Put extra object to check list-by-prefix filters it out
+    temporary_buffer<char> data = sstring("1234567890").release();
+    client->put_object(format("/{}/extra-{}", bucket, ::getpid()), std::move(data)).get();
+    for (int i = 0; i < 12; i++) {
+        temporary_buffer<char> data = sstring("1234567890").release();
+        auto name = format("{}obj.{}", prefix, i);
+        client->put_object(format("/{}/{}", bucket, name), std::move(data)).get();
```
Comment (tchaikov) [inline test/boost/s3_test.cc:396]: `data` has been moved away when putting the extra object and when putting the previous object with `f"testprefix-{pid}"` prefix. so we are actually putting 12 empty objects with the specified prefix.  to minimize surprise, probably we could put a duplicated payload? like:
```c++
client->put_object(format("/{}/extra-{}", bucket, ::getpid()), data.share()).get();
// ...
client->put_object(format("/{}/{}", bucket, name), data.share()).get();
```

or, in this case, we don't really care about the content of the objects, we could just use an empty payload.

**Review by tchaikov — COMMENTED** [review]
left a few comments.

File: test/boost/s3_test.cc (hunk 2 above)
Comment (xemul) [reply test/boost/s3_test.cc:396]: > so we are actually putting 12 empty objects with the specified prefix.

Why? Inside this loop there's another variable that shadows the outer one, no?

> or, in this case, we don't really care about the content of the objects, we could just use an empty payload.

Yes, we don't care the content

File: utils/s3/client.cc [hunk 3]
```diff
@@ -1226,4 +1226,106 @@ future<> client::close() {
...
+}
+
+static sstring parse_list_of_objects(sstring body, std::vector<sstring>& names) {
+    auto doc = std::make_unique<rapidxml::xml_document<>>();
+    try {
+        doc->parse<0>(body.data());
+
+        auto root_node = doc->first_node("ListBucketResult");
+        for (auto contents = root_node->first_node("Contents"); contents; contents = contents->next_sibling()) {
+            auto key = contents->first_node("Key");
```
[the remark this answers is not in the document]
Comment (xemul) [reply utils/s3/client.cc:1245]: I made it symmetrical to parse_tagging(). Do you suggest to fix that part too?

File: utils/s3/client.cc [hunk 4]
```diff
@@ -1226,4 +1226,106 @@ future<> client::close() {
...
+        }
+
+        auto is_truncated = root_node->first_node("IsTruncated");
+        if (std::string_view(is_truncated->value()) == "false") {
+            return "";
+        } else {
+            auto continuation = root_node->first_node("NextContinuationToken");
+            return continuation->value();
+        }
+    } catch (...) {
```
[the remark this answers is not in the document]
Comment (xemul) [reply utils/s3/client.cc:1256]: Ah, you're right. This catching is not required at all. It was supposed to comply with map_s3_client_exception(), but there's outer catch that takes care of it

File: utils/s3/client.cc [hunk 5]
```diff
@@ -1226,4 +1226,106 @@ future<> client::close() {
...
+}
+
+future<> client::bucket_lister::start_listing() {
+    // This is the implementation of paged ListObjectsV2 API call
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html
+    sstring continuation_token;
+    do {
+        s3l.trace("GET /?list-type=2 (prefix={})", _prefix);
+        auto req = http::request::make("GET", _client->_host, format("/{}", _bucket));
+        req.query_parameters["list-type"] = "2";
```
[the remark this answers is not in the document]
Comment (xemul) [reply utils/s3/client.cc:1269]: This way of assigning req.query_parameters is used pretty much everywhere

```
$ git grep 'req.query_parameters' ./utils/s3/client.cc
utils/s3/client.cc:132:    for (const auto& q : req.query_parameters) {
utils/s3/client.cc:323:    req.query_parameters["tagging"] = "";
utils/s3/client.cc:346:    req.query_parameters["tagging"] = "";
utils/s3/client.cc:370:    req.query_parameters["tagging"] = "";
utils/s3/client.cc:627:    req.query_parameters["partNumber"] = format("{}", part_number + 1);
utils/s3/client.cc:628:    req.query_parameters["uploadId"] = _upload_id;
utils/s3/client.cc:676:    req.query_parameters["uploadId"] = std::exchange(_upload_id, ""); // now upload_started() returns false
utils/s3/client.cc:692:    req.query_parameters["uploadId"] = _upload_id;
utils/s3/client.cc:771:    req.query_parameters["partNumber"] = format("{}", part_number + 1);
utils/s3/client.cc:772:    req.query_parameters["uploadId"] = _upload_id;
utils/s3/client.cc:901:        req.query_parameters["uploads"] = "";
utils/s3/client.cc:920:        req.query_parameters["uploadId"] = _upload_id;
utils/s3/client.cc:934:        req.query_parameters["uploadId"] = std::exchange(_upload_id, "");
utils/s3/client.cc:971:        req.query_parameters.emplace("partNumber", to_sstring(part_number + 1));
utils/s3/client.cc:972:        req.query_parameters.emplace("uploadId", _upload_id);
utils/s3/client.cc:1269:        req.query_parameters["list-type"] = "2";
utils/s3/client.cc:1270:        req.query_parameters["max-keys"] = _max_keys;
utils/s3/client.cc:1272:            req.query_parameters["continuation-token"] = std::exchange(continuation_token, "");
utils/s3/client.cc:1275:            req.query_parameters["prefix"] = _prefix;
```

It's probably worth fixing uniformly

File: test/boost/s3_test.cc (hunk 2 above)
Comment (tchaikov) [reply test/boost/s3_test.cc:396]: > > so we are actually putting 12 empty objects with the specified prefix.
> 
> Why? Inside this loop there's another variable that shadows the outer one, no?

ahh, indeed. i didn't realize that you are reusing the same variable name. sorry for the noise.

File: utils/s3/client.cc [hunk 6]
```diff
@@ -1226,4 +1226,106 @@ future<> client::close() {
...
+client::bucket_lister::bucket_lister(shared_ptr<client> client, sstring bucket, sstring prefix, size_t maximum_objects_per_page)
+    : _client(std::move(client))
+    , _bucket(std::move(bucket))
+    , _prefix(std::move(prefix))
+    , _max_keys(format("{}", maximum_objects_per_page))
+    , _queue(512 / sizeof(std::optional<directory_entry>))
+{
+}
+
+static sstring parse_list_of_objects(sstring body, std::vector<sstring>& names) {
```
[the remark this answers is not in the document]
Comment (xemul) [reply utils/s3/client.cc:1238]: I can't say it looks much nicer, but I'm not big fan of returning result via argument either. Will fix

File: utils/s3/client.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply utils/s3/client.cc:1245]: either way, i can send a fix for `parse_tagging()` . as it is not orthogonal to this change.

File: utils/s3/client.cc [hunk 7]
```diff
@@ -1226,4 +1226,106 @@ future<> client::close() {
...
+        doc->parse<0>(body.data());
+
+        auto root_node = doc->first_node("ListBucketResult");
+        for (auto contents = root_node->first_node("Contents"); contents; contents = contents->next_sibling()) {
+            auto key = contents->first_node("Key");
+            names.push_back(key->value());
+        }
+
+        auto is_truncated = root_node->first_node("IsTruncated");
+        if (std::string_view(is_truncated->value()) == "false") {
```
[the remark this answers is not in the document]
Comment (xemul) [reply utils/s3/client.cc:1250]: Refactored this part likewise

**Comment by xemul** [comment]
upd:
- fixed review comments (thanks, Kefu)

**Review by tchaikov — APPROVED** [review]
lgtm

**Comment by xemul** [comment]
upd:
- mark bucket_lister::close() noexcept

File: utils/s3/client.cc [hunk 8]
```diff
@@ -1226,4 +1226,113 @@ future<> client::close() {
...
+                    continuation_token = std::move(list.second);
+                }
+            );
+        } catch (...) {
+            _queue.abort(std::current_exception());
+            co_return;
+        }
+
+        for (auto&& o : names) {
+            co_await _queue.push_eventually(directory_entry{std::move(o)});
```
Comment (avikivity) [inline utils/s3/client.cc:1300]: Could you have used generator and co_yield? I guess cleanup is difficult there.

Comment (tchaikov) [reply utils/s3/client.cc:1300]: seastar's async generator is buggy. a fix is pending on review. see https://github.com/scylladb/seastar/pull/2218

Comment (avikivity) [reply utils/s3/client.cc:1300]: Yeah. It's okay to use a queue meanwhile.

**Comment by xemul** [comment]
#20116

**Comment by xemul** [comment]
upd:
- added test for lister closed early

File: test/boost/s3_test.cc [hunk 9]
```diff
@@ -377,3 +378,33 @@ SEASTAR_THREAD_TEST_CASE(test_client_put_get_tagging) {
...
+    auto close_lister = deferred_close(lister);
+
+    while (auto de = lister.get().get()) {
+        testlog.info("-> [{}]", de->name);
+        auto it = names.find(de->name);
+        BOOST_REQUIRE(it != names.end());
+        names.erase(it);
+    }
+    BOOST_REQUIRE(names.empty());
+}
```
[the remark this answers is not in the document]
Comment (xemul) [reply test/boost/s3_test.cc:417]: Added in v2

File: utils/s3/client.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply utils/s3/client.cc]: created #20160

**Comment by xemul** [comment]
@scylladb/scylla-maint , please consider merging

**Pull request merged**